### PR TITLE
Cleanup Exists Balancers

### PR DESCRIPTION
**Pull request by maxsu**
Effort to clean and refactor existsBalancers / addSplitterToFlow

Work So Far:
* Start using clang-format to tidy code
* Extract addSplitter to network_tools:addSplitterToFlow
* Use meaningful variables (not n or m)
* Extract sortMatrix function
* Write rowConcat, and constRow function
* Write some wiring utility functions
  * isWired tells us if a wiring integer is > -1 (wired) or not
  * wiredConnections gives us the wired connections in a wiring

* Simplify `splitter_flow` generation
  * We only iterate over the splitters wired inputs and outputs
  * We extract a loop_correction factor
  * We only do one row multiply
  * We use constRow and rowConcat to build the final splitter_flow
  * splitter_flow is now a const

* Simplify the flow row extension loop
  * We only extend the  flow rows in one place
  * We drop the added_inputs boolean flag

### CODE REVIEW

**Comment by maxsu** [comment]
Hi @exfret,

This PR includes the work from your PRs, and pushes some more work. This is still a draft

Please review the commits and let me know if you have suggestions for more changes!

I didn't include your Unit test changes, but I'm whipping up the Catch2 unit tests this evening

**Comment by maxsu** [comment]
Actually, would you like to test reviewing and approving changes?

As you go through each commit, click the review button. There you can either approve the commit, or comment about changes that need to happen!

**Review by exfret — APPROVED** [review]

**Review by exfret — APPROVED** [review]

**Review by exfret — APPROVED** [review]

File: lib/network_tools.cpp [hunk 1]
```diff
@@ -127,35 +127,40 @@ using Config = vector<Wiring>;
...
   Row new_splitter_flow = splitter_flow;
-  for (int i = 0; i < m; ++i) {
+  for (int i = 0; i < num_splitter_outputs; ++i) {
     if (splitter_outputs[i] != -1) {
-      for (int j = 0; j < n + M; ++j) {
+      for (int j = 0; j < num_flow_inputs + num_splitter_inputs; ++j) {
         new_splitter_flow[j] *= 1 / (1 - splitter_flow[splitter_outputs[i]]);
+
+        if (splitter_flow[splitter_outputs[i]] == 1) {
+          assert(1 == 0);  // Verify this branch never runs
```
Comment (exfret) [inline lib/network_tools.cpp]: Good idea adding this assert. Kept this in by accident for debugging, but checking that we don't have NaN's creep in would be a good thing to do.

**Review by exfret — APPROVED** [review]

File: lib/network_tools.cpp [hunk 2]
```diff
@@ -94,6 +94,25 @@ Matrix transpose(Matrix matrix) {
   return transpose_matrix;
 }
 
+inline void sortMatrix(Matrix& matrix) {
```
Comment (maxsu) [inline lib/network_tools.cpp:104]: Get rid of inline

**Review by exfret — APPROVED** [review]
Make sure to get rid of inline's - functions should not be inline a the moment.

**Review by exfret — APPROVED** [review]
Discussed how to store "unwired connections" (currently they're stored as negative ones). In particular, discussed wiredConnections() function. Decided to keep this function.

**Review by exfret — APPROVED** [review]

File: lib/network_tools.cpp [hunk 3]
```diff
@@ -170,42 +170,27 @@ Matrix addSplitterToFlow(Matrix flow, const Wiring splitter_inputs,
...
-          splitter_flow = {};
-        }
-      }
-
-      new_splitter_flow[splitter_outputs[i]] = 0;
-    }
+  // Remove circular dependencies from new outputs to old inputs
+  double loop_correction = output_fraction;
+  for (int output : wired_outputs) {
+    loop_correction /= 1 - input_flow[output] * output_fraction;
```
Comment (exfret) [inline lib/network_tools.cpp:187]: Assert could be written here to check that input_flow[output]*output_fraction is in the interval [0,1) (i.e.- greater than or equal to zero and less than 1).

File: lib/network_tools.cpp [hunk 4]
```diff
@@ -170,42 +170,27 @@ Matrix addSplitterToFlow(Matrix flow, const Wiring splitter_inputs,
...
-        if (splitter_flow[splitter_outputs[i]] == 1) {
-          assert(1 == 0);  // Verify this branch never runs
-          splitter_flow = {};
-        }
-      }
-
-      new_splitter_flow[splitter_outputs[i]] = 0;
-    }
+  // Remove circular dependencies from new outputs to old inputs
+  double loop_correction = output_fraction;
```
Comment (maxsu) [inline lib/network_tools.cpp:185]: Move output_fraction to its own thing, either before this loop or later with the row multiplication

File: lib/network_tools.cpp [hunk 5]
```diff
@@ -170,42 +170,27 @@ Matrix addSplitterToFlow(Matrix flow, const Wiring splitter_inputs,
...
   // Find this splitter's output in terms of its input flows
-  for (int i = 0; i < num_splitter_inputs; ++i) {
-    if (splitter_inputs[i] != -1) {
-      // Flow to splitter along its i'th input
-      Row input_flow = flow[splitter_inputs[i]];
-      for (int j = 0; j < num_flow_inputs; ++j) {
-        splitter_flow[j] += input_flow[j] / num_splitter_outputs;
-      }
-    }
+  Row input_flow = zeroRow(num_flow_inputs);
```
Comment (exfret) [inline lib/network_tools.cpp:179]: Consider renaming this to something that mentions its relationship to the splitter, like flowToSplitterFromInputs.

**Review by exfret — APPROVED** [review]

**Pull request merged**